### DISCUSSION TITLE
Unable to import Geometry into spaceclaim

### DISCUSSION THREAD

**Discussion by Leokek1** [ansys/pyansys-geometry#1430] [Q&A] [answered]
Trying to import geometry to ansys, however whenever I run launch_modeler_with_spaceclaim I get the following error:
Code Inputted: 

```
import os
import ansys.fluent.core as pyfluent
from ansys.geometry.core import Modeler
from ansys.geometry.core import launch_modeler_with_spaceclaim

# modeler = Modeler()

# Launch the modeler
modeler = launch_modeler_with_spaceclaim() #script doesn't get past this point

# Import geometry
design = modeler.open_file(
    os.path.join(
        os.getcwd(),
        f"003659.scdocx",
    )
)


# Retrieve the geometry
injector = design.bodies[0]

# Display the part
design.plot()'

```
Error Received: 

```
'C:\Users\user.aad\OneDrive - Personal Project\General - PROJECT 21\Engineering\Analysis\Automated WindmillSimulations>C:/Users/user.aad/AppData/Local/Microsoft/WindowsApps/python3.11.exe "c:/Users/user.aad/OneDrive - Personal Project/General - PROJECT 21/Engineering/Analysis/Automated WindmillSimulations/named_selections.py"
CRITICAL -  -  logger - handle_exception - Uncaught exception
Traceback (most recent call last):
  File "c:\Users\user.aad\OneDrive - Personal Project\General - PROJECT 21\Engineering\Analysis\Automated WindmillSimulations\named_selections.py", line 9, in <module>
    modeler = launch_modeler_with_spaceclaim()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\user.aad\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\ansys\geometry\core\misc\checks.py", line 403, in wrapper
    return method(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\user.aad\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\ansys\geometry\core\connection\launcher.py", line 855, in launch_modeler_with_spaceclaim
    return prepare_and_start_backend(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\user.aad\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\ansys\geometry\core\connection\product_instance.py", line 346, in prepare_and_start_backend
    return Modeler(
           ^^^^^^^^
  File "C:\Users\user.aad\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\ansys\geometry\core\modeler.py", line 106, in __init__
    self._grpc_client = GrpcClient(
                        ^^^^^^^^^^^
  File "<@beartype(ansys.geometry.core.connection.client.GrpcClient.__init__) at 0x16401c49ee0>", line 236, in __init__
  File "C:\Users\user.aad\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\ansys\geometry\core\connection\client.py", line 172, in __init__
    grpc_backend_response = self._admin_stub.GetBackend(Empty())
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\user.aad\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\grpc\_channel.py", line 1181, in __call__
    return _end_unary_response_blocking(state, call, False, None)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\user.aad\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\grpc\_channel.py", line 1006, in _end_unary_response_blocking
    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.UNIMPLEMENTED
        details = ""
        debug_error_string = "UNKNOWN:Error received from peer  {grpc_message:"", grpc_status:12, created_time:"2024-09-16T16:37:14.9870495+00:00"}"'

```

**Comment by RobPasMue** [comment]
Hi @Leokek1 - thanks for opening this issue. I have several questions for you:

* Can you share your python environment with us? (i.e. output from ``pip list``)
* Which SpaceClaim version do you have installed?
* Has the script ever worked?
* Does SpaceClaim's GUI show up even though the error message appears?

**Reply by Leokek1** [reply]
1. Output:
```
Package                               Version     Editable project location
------------------------------------- ----------- ------------------------------------------------------------------------------------------------------------------------------------------------------
about-time                            4.2.1
aiohappyeyeballs                      2.4.0
aiohttp                               3.10.5
aiosignal                             1.3.1
alive-progress                        3.1.5
ansys-api-dbu                         0.3.4
ansys-api-fluent                      0.3.27
ansys-api-geometry                    0.4.8
ansys-api-platform-instancemanagement 1.1.0
ansys-api-tools-filetransfer          0.1.0
ansys-fluent-core                     0.27.dev0   C:\Users\david.liljewall.aad\OneDrive - Agile Space Industries Inc\General - A2200\Engineering\Analysis\Automated Pintle Injector Simulations\pyfluent
ansys-geometry-core                   0.7.2
ansys-platform-instancemanagement     1.1.2
ansys-tools-filetransfer              0.1.0
ansys-tools-path                      0.6.0
ansys-tools-visualization-interface   0.4.4
ansys-units                           0.3.2
appdirs                               1.4.4
asttokens                             2.4.1
async-lru                             2.0.3
attrs                                 23.2.0
autopep8                              2.0.4
Babel                                 2.12.1
backcall                              0.2.0
beartype                              0.18.5
beautifulsoup4                        4.12.3
black                                 23.10.1
bleach                                6.0.0
bs4                                   0.0.2
certifi                               2024.7.4
cffi                                  1.16.0
charset-normalizer                    3.3.2
click                                 8.1.7
clr-loader                            0.2.6
colorama                              0.4.6
comm                                  0.2.2
contourpy                             1.1.0
cycler                                0.11.0
debugpy                               1.8.2
decorator                             5.1.1
defusedxml                            0.7.1
docker                                7.1.0
et-xmlfile                            1.1.0
exceptiongroup                        1.2.2
executing                             2.0.1
fastjsonschema                        2.18.0
filelock                              3.12.2
flexcache                             0.3
flexparser                            0.3.1
fonttools                             4.41.1
fqdn                                  1.5.1
frozenlist                            1.4.1
grapheme                              0.6.0
grpcio                                1.66.1
grpcio-health-checking                1.62.3
h11                                   0.14.0
idna                                  3.7
importlib_metadata                    8.5.0
ipykernel                             6.29.5
ipython                               8.26.0
ipython-genutils                      0.2.0
IronPdf                               2024.7.1.8
jedi                                  0.19.1
Jinja2                                3.1.4
joblib                                1.4.2
json5                                 0.9.14
jsonpointer                           2.4
jupyter_client                        8.6.2
jupyter_core                          5.7.2
jupyterlab-pygments                   0.2.2
jupyterlab-widgets                    3.0.8
kiwisolver                            1.4.4
LabJackPython                         2.1.0
lxml                                  5.3.0
MarkupSafe                            2.1.5
matplotlib                            3.7.2
matplotlib-inline                     0.1.7
mistune                               3.0.1
molmass                               2023.8.30
more-itertools                        10.5.0
mpmath                                1.3.0
msgpack                               1.1.0
multidict                             6.1.0
mypy-extensions                       1.0.0
nest-asyncio                          1.6.0
networkx                              3.1
nltk                                  3.9.1
npTDMS                                1.7.1
numpy                                 1.25.1
openpyxl                              3.1.2
outcome                               1.3.0.post0
overrides                             7.3.1
packaging                             24.1
pandas                                2.0.3
pandocfilters                         1.5.0
parso                                 0.8.4
pathspec                              0.11.2
pdfkit                                1.0.0
pexpect                               4.9.0
pickleshare                           0.7.5
Pillow                                10.0.0
Pint                                  0.24.3
pip                                   24.2
platformdirs                          4.2.2
pooch                                 1.8.2
prometheus-client                     0.17.1
prompt_toolkit                        3.0.47
protobuf                              4.25.4
psutil                                6.0.0
ptyprocess                            0.7.0
pure_eval                             0.2.3
pycodestyle                           2.11.1
pycparser                             2.21
Pygments                              2.18.0
pyhtml2pdf                            0.0.7
pyparsing                             3.0.9
PySocks                               1.7.1
python-dateutil                       2.9.0.post0
python-dotenv                         1.0.1
python-json-logger                    2.0.7
pythonnet                             3.0.3
pytz                                  2023.3
pyvista                               0.44.1
pywin32                               306
pywinpty                              2.0.11
PyYAML                                6.0.1
pyzmq                                 26.0.3
QtPy                                  2.3.1
referencing                           0.30.0
regex                                 2024.9.11
requests                              2.32.3
rfc3339-validator                     0.1.4
rfc3986-validator                     0.1.1
rocketprops                           0.1.7
rpds-py                               0.9.2
scipy                                 1.11.1
scooby                                0.10.0
selenium                              4.23.1
semver                                3.0.2
Send2Trash                            1.8.2
six                                   1.16.0
sniffio                               1.3.0
sortedcontainers                      2.4.0
soupsieve                             2.5
stack-data                            0.6.3
sympy                                 1.12
terminado                             0.17.1
Theano                                1.0.5
tinycss2                              1.2.1
tornado                               6.4.1
tqdm                                  4.66.5
traitlets                             5.14.3
trame                                 3.6.5
trame-client                          3.2.5
trame-server                          3.1.2
trame-vtk                             2.8.10
trame-vuetify                         2.7.1
trio                                  0.26.0
trio-websocket                        0.11.1
typing_extensions                     4.12.2
tzdata                                2023.3
uri-template                          1.3.0
urllib3                               2.2.2
vtk                                   9.3.1
wcwidth                               0.2.13
webcolors                             1.13
webdriver-manager                     4.0.1
webencodings                          0.5.1
websocket-client                      1.8.0
websockets                            13.0.1
widgetsnbextension                    4.0.8
wkhtmltopdf                           0.2
wslink                                2.1.3
wsproto                               1.2.0
yarl                                  1.11.1
zipp                                  3.20.2

```
2. ANSYS Version is all 2023 R2 
3. No the script has never worked
4. The SpaceClaim GUI has shown up even though the error message appears

Thank you so much for the help and the really fast answers!

**Reply by RobPasMue** [reply]
Hi @Leokek1 - point 2 answers my doubts! =) As you can see here: https://geometry.docs.pyansys.com/version/stable/getting_started/compatibility.html compatibility for Ansys 23R2 is very limited. You would only be able to use ``ansys-geometry-core`` versions ``0.3.x``. From ``0.4.x`` onwards, PyAnsys Geometry is backwards compatible (starting on 24R1).

This means that you will be able to use any PyAnsys Geometry version from 0.4 onwards against Ansys 24R1, 24R2, 25R1...
However, for 23R2 (our first actual release) we had to deprecate support very early due to many problems for evolving the library otherwise. The API of PyAnsys Geometry for version 0.3 has many changes than the actual one, but you can still have access to its documentation here: https://geometry.docs.pyansys.com/version/0.3/

Would it be possible for you to update to a newer Ansys release? (24R1 or 24R2?)

**Reply by Leokek1** [reply]
I won't be able to Update to a newer Ansys release, is there any other work around for this? Spaceclaim does open, it just doesn't load in my part.

**Reply by Leokek1** [reply]
I tried doing some debugging and I got SpaceClaim to import the file but now I am getting some new errors. This happens once the file is imported into SC. I do have an instance of Fluent running and a instance of WorkBench open:
Input Code:
```
import pkg_resources

# Enforce the recommended anays-geometry-core version to be compatible with ANSYS 2023R2
pkg_resources.require("ansys-geometry-core==0.3.2")

from ansys.geometry.core import launch_modeler_with_spaceclaim

modeler = launch_modeler_with_spaceclaim()

modeler.open_file(file_path=rf"[insert file path here'")

```

It then throws this following error:
![image](https://github.com/user-attachments/assets/144bbb78-6d0e-4ae9-acec-952f12c31231)

**Reply by RobPasMue** [reply]
Hi @Leokek1 

Based on your code I am assuming you are using Ansys 23R2 and ``ansys-geometry-core==0.3.2``.

I am also guessing until here everything works fine:

```py
import pkg_resources

# Enforce the recommended anays-geometry-core version to be compatible with ANSYS 2023R2
pkg_resources.require("ansys-geometry-core==0.3.2")

from ansys.geometry.core import launch_modeler_with_spaceclaim

modeler = launch_modeler_with_spaceclaim()
```

And then you are getting the error when running ``modeler.open_file(file_path=rf"[insert file path here}")`` - for your version ``0.3.2`` you can see the API docs here: https://geometry.docs.pyansys.com/version/0.3/api/ansys/geometry/core/modeler/Modeler.html#Modeler.open_file

I would recommend that you call the function with an additional argument: ``modeler.open_file(file_path=rf"[insert file path here}", upload_to_server=False)``. Since you are using SpaceClaim as your backend and it is running in your local machine you don't need to upload. I also recall some issues in the past with 23R2 and the upload functionality... give it a try and let me know! :)

**Reply by Leokek1** [reply]
Just gave that a try and it still returns the same exact error unfortunately!
![image](https://github.com/user-attachments/assets/b693310c-3bd7-4b8b-85a5-46630b03627c)
 Here is the image of the error for reference.

**Reply by RobPasMue** [reply, marked as answer]
That's sad to read... We dropped support early on the 23R2 version because we hit a lot of problems. The first stable release starts with 24R1. 

You can see some older discussions we had here: https://github.com/ansys/pyansys-geometry/discussions/754#discussioncomment-7082980 - on 23R2 we did "the best we could" with SpaceClaim and Discovery but it surely wasn't optimal. Do you have access to the Ansys Geometry Service? (i.e. is there a folder on you machine with the following path? ``C:\Program Files\ANSYS Inc\v232\GeometryService``)

If so, I'd suggest you use the function ``launch_modeler_with_geometry_service`` and gave it a try as well.

**Reply by Leokek1** [reply]
Hi! Sorry for not following up, however gave it a go and it also has that problem. I was thinking of waiting to upgrade to the new ANSYS 24R1, but once I do I will try it once more and hopefully it won't have this issue. I did almost find some work arounds though in case anyone reads this in the future and wants to follow the lead. Spaceclaim has scripting capabilities with both block and scripts. Script editing allows the opening, naming, and everything that is desired. However, block editing gets AUTO-turned on as soon as a part is opened. Since both of these options can't take place at the same time and block editing has precedence, it just turns off the script. This issue is basically what I ran into, if someone figures out how to turn off the auto block editing start please do let me know! I would love to find out.

**Reply by RobPasMue** [reply]
HI @b-matteo - do you know if it is possible to do what @Leokek1 is asking for inside SpaceClaim directly?

**Reply by b-matteo** [reply]
Hi @Leokek1, @RobPasMue thanks for reaching out.
I think you can get the script to work:
- Open your document into Ansys SpaceClaim
- In the Design tab, disable the _Blocks_ function.
- Save your changes

from now on, you shouldn't get the _Blocks_ editing turned on when you open the file.

Please let us know if that worked.

Thanks

**Reply by Leokek1** [reply]
Hi @b-matteo, so I gave that a whirl. However, part of my script is to import a part. So whenever I import the part block recording gets auto-enabled in the NEW spaceclaim instance. So once I do deactivate block recording, that instance does work -- however, once I import the part (the first line of my script) it proceeds to just not work since both are activated and block recording takes precedence. If I don't have the import statement, import in the part myself and run the rest of the script, it works beautifully. Its just the automatic enabling of the block recording which stops it by importing it via the spaceclaim script.
![image](https://github.com/user-attachments/assets/ac3698c3-e5b5-4d0c-89a6-b7a137cb0944)

**Reply by Leokek1** [reply]
@RobPasMue @b-matteo 
When I load in the Solidpart to SpaceClaim it does load in. It just gives me that error. Would it be possible to catch the exception and then relink python with the active instance of Spaceclaim (since the file stays open and is perfect, but the launch_modeler_with_spaceclaim just auto-breaks the code)? I was thinking of using win32com but don't know the prog ID to use that. From there I was also wondering, how do you "activate" a spaceclaim script in spaceclaim if you happen to know that (using the original Python code). Thanks so much!

**Reply by Leokek1** [reply]
Okay, so I ended up getting ANSYS v24R2 and now it does work! I have some further questions if that's okay? I was wondering if there were a PYANSYS geometry command to run a spaceclaim script once I open a launch_modeler_with_spaceclaim(product_version= 242 instance? I have a spaceclaim script which does everything I need and I tried to translate the named selection features to PYAnsys Geometry, but I can't find the right functions/classes -- IDesignBody, GetRootPart(), and a few others I can't find. I know that Spaceclaim does use python, i just don't know the imports that spaceclaim has/uses for my script to fully work.

**Reply by Leokek1** [reply]
Actually got most of it to work! The only thing I have left to ask is if there is an "explode" named selection python method for my named selections. I have been sifting through the documentation and was wondering if that were a method I could use! If not, is there any workaround.. e.g. accessing the list of faces and then exploding it manually in my code? Thanks!

**Reply by RobPasMue** [reply]
Hi @Leokek1! We are able of creating named selections directly using the PyAnsys Geometry API - have a look at https://geometry.docs.pyansys.com/version/stable/api/ansys/geometry/core/designer/design/Design.html#Design.create_named_selection

There is an example on how to do this here: https://geometry.docs.pyansys.com/version/stable/examples/04_applied/02_naca_fluent.html#Create-named-selections

Nonetheless, there is also the possibility to run a SpaceClaim script if that is your preferred option. We have the following API for that purpose: https://geometry.docs.pyansys.com/version/stable/api/ansys/geometry/core/modeler/Modeler.html#Modeler.run_discovery_script_file

Also, a "Design" provides you access to the list of bodies, a "Body" object provides you access to the "Face" objects, and the "Face" object provides you access to the "Edge" objects. However, NamedSelections do not contain a list of the elements they compose. If the NamedSelection was created by PyAnsys Geometry it'd be easy to determine, but if the Named Selection was created on the service, from the Python client it is currently not possible. We are working on it (on the products) to return the elements that compose a NamedSelection, but this is not available yet. We are tracking it here: https://github.com/ansys/pyansys-geometry/issues/1101

Hope all this helps!

**Reply by Leokek1** [reply]
It totally does! I ended up just using the PyAnsys Geometry service and did it that way! I just finished the brunt of the coding and debugging and will not be integrating all of it, super excited. Thank you so much for all the help you have given!

**Reply by RobPasMue** [reply]
Awesome, glad it works @Leokek1!

**Reply by RobPasMue** [reply]
Also! If you like the project, we'd appreciate your support with a star (see "star" on landing page of https://github.com/ansys/pyansys-geometry) =) that will give us some visibility in the open source world. Thanks in advance as well for your time and we're here for more doubts you may have!
